### ISSUE TITLE
improve error message for missing return type

### ISSUE THREAD

**Issue by emekoi** [ziglang/zig#1672]
when the return type is missing for a function, i get the error `expected token '{', found 'fn'`. could someone point me towards the code that handles parsing of functions, so i could fix this?

**Comment by thejoshwolfe** [comment]
I actually think we should disallow return type expressions starting with `{`. That would solve this problem.

Here's the real question: why would you ever want a return expression to legitimately begin with `{`? I believe it's never going to work semantically, since the only value a `{...}` block could possibly evaluate to is the `void` value `{}`. You would have to put a label on the block to get any other value, and in that case the expression doesn't start with `{` anymore.

**Comment by dbandstra** [comment]
I would make this a priority... It's something that newcomers are going to run into again and again. (As well as people who frequently switch between Zig and another language like JS.) The error message still initially confuses me after the tenth time running into it.

And if there was a reason to allow return types to start with `{`,  I think it would be worth adding a new symbol to the syntax introducing the return type.

For me, this issue ranks right up there with the 'copying fieldParentPtr thing out of a struct' footgun.

**Comment by winksaville** [comment]
I too run frequently forget the return type and stumble into this all the
time. So I look forward to seeing a better error message.

**Comment by emekoi** [comment]
@thejoshwolfe i think we're talking about two different things here. i'm talking about the error message you get from code like this:
```
fn foo() {

}

fn bar() void {

}
```

**Comment by Hejsil** [comment]
This is related to #760, and I'm currently working on a [solution](https://github.com/ziglang/zig/issues/760#issuecomment-430938743).

**Comment by Hejsil** [comment]
After #1685 we now get this error:
```
/home/jhc/repo/zig/test.zig:1:8: error: invalid token: '{'
fn a() {}
       ^
```

Not perfect, but we get the error earlier, which should help alot.

**Comment by andrewrk** [comment]
I would consider this solved, with a low priority further improvement being to special case this error and add something like "expected return type"

**Comment by nektro** [comment]
with Zig master this now works as expected

```zig
pub fn main() {
}
```

```
test.zig:1:15: error: expected return type expression, found '{'
pub fn main() {
              ^
```
